controllers/openstackcluster_controller.go

```diff
@@ -103,15 +99,66 @@ func (r *OpenStackClusterReconciler) Reconcile(request ctrl.Request) (_ ctrl.Res
 
 	// Handle deleted clusters
 	if !openStackCluster.DeletionTimestamp.IsZero() {
-		return r.reconcileClusterDelete(logger, cluster, openStackCluster)
+		return reconcileDelete(log, r.Client, cluster, openStackCluster)
 	}
 
 	// Handle non-deleted clusters
-	return r.reconcileCluster(ctx, logger, patchHelper, cluster, openStackCluster)
+	return reconcileNormal(ctx, log, r.Client, patchHelper, cluster, openStackCluster)
+}
+
+func reconcileDelete(log logr.Logger, client client.Client, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster) (ctrl.Result, error) {
+	log.Info("Reconciling Cluster delete")
+
+	clusterName := fmt.Sprintf("%s-%s", cluster.Namespace, cluster.Name)
+	osProviderClient, clientOpts, err := provider.NewClientFromCluster(client, openStackCluster)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	networkingService, err := networking.NewService(osProviderClient, clientOpts, log)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	loadBalancerService, err := loadbalancer.NewService(osProviderClient, clientOpts, log, openStackCluster.Spec.UseOctavia)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	if openStackCluster.Spec.ManagedAPIServerLoadBalancer {
+		err = loadBalancerService.DeleteLoadBalancer(clusterName, openStackCluster)
+		if err != nil {
+			return reconcile.Result{}, errors.Errorf("failed to delete load balancer: %v", err)
+		}
+	}
+
+	// Delete other things
+	if openStackCluster.Status.GlobalSecurityGroup != nil {
+		log.Info("Deleting global security group", "name", openStackCluster.Status.GlobalSecurityGroup.Name)
+		err := networkingService.DeleteSecurityGroups(openStackCluster.Status.GlobalSecurityGroup)
+		if err != nil {
+			return reconcile.Result{}, errors.Errorf("failed to delete security group: %v", err)
+		}
+	}
+
+	if openStackCluster.Status.ControlPlaneSecurityGroup != nil {
+		log.Info("Deleting control plane security group", "name", openStackCluster.Status.ControlPlaneSecurityGroup.Name)
+		err := networkingService.DeleteSecurityGroups(openStackCluster.Status.ControlPlaneSecurityGroup)
+		if err != nil {
+			return reconcile.Result{}, errors.Errorf("failed to delete security group: %v", err)
+		}
+	}
+
+	// TODO(sbueringer) Delete network/subnet/router/... if created by CAPO
+
+	// Cluster is deleted so remove the finalizer.
+	controllerutil.RemoveFinalizer(openStackCluster, infrav1.ClusterFinalizer)
+
+	return reconcile.Result{}, nil
 }
 
-func (r *OpenStackClusterReconciler) reconcileCluster(ctx context.Context, logger logr.Logger, patchHelper *patch.Helper, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster) (_ ctrl.Result, reterr error) {
-	logger.Info("Reconciling Cluster")
+func reconcileNormal(ctx context.Context, log logr.Logger, client client.Client, patchHelper *patch.Helper, cluster *clusterv1.Cluster, openStackCluster *infrav1.OpenStackCluster) (ctrl.Result, error) {
```

Comment:
Any particular reason to change this from a method on OpenstackClusterReconciler to a separate function?

Reply:
Mostly to make it consistent with CAPA for an easier diff :). But then later on I saw that CAPA itself does it differently between cluster and machine controller and I lost track of it a bit. I rolled it back.